### ISSUE TITLE
Allow configuration of allow_chunked_length

### ISSUE THREAD

**Issue by moderation** [projectcontour/contour#3221]
We are seeing users getting EOF errors when trying to pull artifacts from a repository hosted behind Contour. These errors are not experienced with an NGINX ingress.

The Artifactory documentation on configuring NGINX recommends enabling chunked transfer encoding - https://www.jfrog.com/confluence/display/RTF6X/Artifactory+Enterprise+Plus#ArtifactoryEnterprisePlus-ModifyNGINXConfiguration

Dropbox shared a recent post about their migration from nginx to envoy https://dropbox.tech/infrastructure/how-we-migrated-dropbox-from-nginx-to-envoy. They explicitly call out the following
> Request that have both Content-Length and Transfer-Encoding: chunked. Requests like that used to work with Nginx, but were broken by Envoy migration.

A new setting to match the NGINX behavior was recently merged into envoy v1.16.0 (from Dropbox) - https://github.com/envoyproxy/envoy/issues/11398 / https://github.com/envoyproxy/envoy/pull/12349/files

Contour 1.10.0 includes Envoy 1.16.0. Envoy documentation for this feature at https://www.envoyproxy.io/docs/envoy/v1.16.0/api-v3/config/core/v3/protocol.proto.html?highlight=allow_chunked_length#config-core-v3-http1protocoloptions

@stevesloka has a fork with this setting enabled in Contour - https://github.com/stevesloka/contour/commit/88b17c268bc18b54c342db5a6170f9695b110389

Request that this functionality be added to Contour. /cc @moondev who contributed to this research.

**Comment by youngnick** [comment]
Spent a little while reading the issues and PRs here, here are a couple of early thoughts:
- enabling this configuration is a (small) security risk - this combination of headers can be used for smuggling data. This doesn't mean we shouldn't do it - there's a clear need here. But it does mean that we shouldn't turn this on by default, or even make it a global config item. It needs to be a per-host or per-route setting. Feels like a boolean will be the go.
- The setting we'll need to use is in [config.core.v3.Http1ProtocolOptions](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#config-core-v3-http1protocoloptions), so I think it will need to be a cluster-level setting on Envoy output.

In short, I agree we need to do something here, it will need to be a HTTPProxy field somewhere, we just need to design and implement it. I'm interested in your thoughts, @moderation and @moondev, on if you think a per-vhost or per-route setting is more useful?

**Comment by moderation** [comment]
I think per-route makes sense for this as it is granular and would be opt-in. For our use case the team could enable for the particular routes serving artifacts.  On the second bullet point I think you are referring to an Envoy upstream cluster? Yes, would make sense to set there I think.

**Comment by stevesloka** [comment]
It looks like this is a `listener` or a `cluster` setting. I think it makes sense to add to clusters, so something like this:

```yaml
apiVersion: projectcontour.io/v1
kind: HTTPProxy
metadata:
  name: name-example-foo
spec:
  virtualhost:
    fqdn: foo1.bar.com
  routes:
    - conditions:
      - prefix: /
      services:
        - name: s1
           port: 80
           allowChunkedLength: true
```

The alternative would be to create an `httpProtocolPolicy` struct to put this into like we've done in the past:
```yaml
apiVersion: projectcontour.io/v1
kind: HTTPProxy
metadata:
  name: name-example-foo
spec:
  virtualhost:
    fqdn: foo1.bar.com
  routes:
    - conditions:
      - prefix: /
      services:
        - name: s1
           port: 80
           httpProtocolPolicy:
             allowChunkedLength: true
```

Do we need this for ingress/v1? If so we'd need to come up with an annotation on the service or something.

**Comment by jpeach** [comment]
I'm not convinced this needs to be a per-route settings. Handling the transfer encoding is purely in the domain of the proxy and is an independent decision from what the origin needs. Reading the RFC, the behavior of this case is very clearly specified, so turning on the Envoy flag simply makes it behave in a standards-conformant manner.

I don't think that users should need to do tweaks to get conformant behavior, though I am sympathetic to the need to tweak to work around broken stuff. That does't see to be the case here though.

My reading of the RFC is that is addresses the smuggling issue:

       If a message is received with both a Transfer-Encoding and a
       Content-Length header field, the Transfer-Encoding overrides the
       Content-Length.  Such a message might indicate an attempt to
       perform request smuggling (Section 9.5) or response splitting
       (Section 9.4) and ought to be handled as an error.  A sender MUST
       remove the received Content-Length field prior to forwarding such
       a message downstream.

"ought to" is not normative, but "MUST remove" is. So the presence of both headers creates a (historical)( ambiguity that the sender must remove before forwarding.

**Comment by youngnick** [comment]
So, @jpeach, your opinion is that we should just enable this flag everywhere, possibly with a config option to disable it?

**Comment by jpeach** [comment]
> So, @jpeach, your opinion is that we should just enable this flag everywhere, possibly with a config option to disable it?

Yeh, my reading of the RFC is that this case is well defined. Request attacks happen when different intermediaries make different decisions about this header. But if Envoy is following the RFC here, there's no ambiguity.

https://github.com/httpwg/http-core/issues/617

**Comment by youngnick** [comment]
Yeah, you make a good point. I'll walk back this one then, and add one to my personal "got it wrong" metric.

I think that we *should* implement this feature by default (since it makes Envoy behave in a more RFC-compliant way), and the security risk is managed by Envoy performing the RFC-compliant behavior and stripping the `Content-Length` header if `Transfer-Encoding: chunked` is set.

I propose also having an emergency off switch in the configuration file that will disable this behavior, with a note to log an issue if you use it, and a reference here.

**Comment by sunjayBhatia** [comment]
Taking this unless there are any objections

**Comment by sunjayBhatia** [comment]
> It looks like this is a `listener` or a `cluster` setting. I think it makes sense to add to clusters, so something like this:
> 
> ```yaml
> apiVersion: projectcontour.io/v1
> kind: HTTPProxy
> metadata:
>   name: name-example-foo
> spec:
>   virtualhost:
>     fqdn: foo1.bar.com
>   routes:
>     - conditions:
>       - prefix: /
>       services:
>         - name: s1
>            port: 80
>            allowChunkedLength: true
> ```
> 
> The alternative would be to create an `httpProtocolPolicy` struct to put this into like we've done in the past:
> 
> ```yaml
> apiVersion: projectcontour.io/v1
> kind: HTTPProxy
> metadata:
>   name: name-example-foo
> spec:
>   virtualhost:
>     fqdn: foo1.bar.com
>   routes:
>     - conditions:
>       - prefix: /
>       services:
>         - name: s1
>            port: 80
>            httpProtocolPolicy:
>              allowChunkedLength: true
> ```
> 
> Do we need this for ingress/v1? If so we'd need to come up with an annotation on the service or something.

Do we need to set this on the upstream clusters? At least for ExtensionService configured clusters it seems we only support h2/h2c and this is a HTTP1 option: https://github.com/projectcontour/contour/blob/0c73316ff31fdb5b5d9eebd0209eab3066ff9859/apis/projectcontour/v1alpha1/extensionservice.go#L79-L84

**Comment by youngnick** [comment]
I think we should just add this everywhere, and have an option to globally disable it if required.  Since Envoy is now enforcing the RFC behavior if this option is enabled, we should just do that, and provide an escape hatch in case it breaks someone.

I'll leave it up to you to decide if we should bother detecting the cluster version, or just set it and let Envoy sort it out.

**Comment by sunjayBhatia** [comment]
Cross posting https://github.com/projectcontour/contour/pull/3248#issuecomment-761108131

In case anyone on this thread has thoughts about reproing this without Artifactory etc.
